jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -463,90 +454,148 @@ def mos_image_parser(app, data_obj, data_labels=None, share_image=0,
 
 
 @data_parser_registry("mosviz-metadata-parser")
-def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False, repeat=1):
+def mos_meta_parser(app, data_obj=None, ids=None):
     """
-    Attempts to parse MOS FITS header metadata.
+    Extracts specific metadata from each data entry's .meta
 
     Parameters
     ----------
     app : `~jdaviz.app.Application`
         The application-level object used to reference the viewers.
     data_obj : str or list or HDUList
-        File path, list, or an HDUList to extract metadata from.
+        Optional fallback arg for Identifier column
+        File path, list, or an HDUList to extract metadata from if
+        metadata isn't found in Data.meta, search for metadata
     ids : list of str
+        Optional fallback arg for Identifier column
         A list with identification strings to be used to label mosviz
-        table rows. Typically, a list with file names.
-    spectra : bool, optional
-        In case the FITS objects are related to spectral data.
-    sp1d : bool, optional
-        Indicates if the input ``data_obj`` is a 1D rather than 2D spectrum.
-    repeat : int, optional
-        If greater than 1, indicates that the input file metadata applies
-        to multiple table rows.
+        table rows. Typically, a list with file names. Used in combination
+        with data_obj for fallback
     """
 
-    if data_obj is None:
-        return
+    with app.data_collection.delay_link_manager_update():
 
-    # Coerce into list if needed
-    if not isinstance(data_obj, (list, tuple, SpectrumCollection)):
-        data_obj = [data_obj]
+        current_columns = [comp.label for comp in app.data_collection['MOS Table'].main_components]
+        # source name can be taken from 1d spectra
+        if "1D Spectra" in current_columns:
+            names = _get_source_identifiers(app, "1D Spectra", data_obj, ids)
+            _add_to_table(app, names, "Identifier")
 
-    data_obj = [fits.open(x) if _check_is_file(x) else x for x in data_obj]
+        # metadata taken from 2d spectra
+        if "2D Spectra" in current_columns:
+            filters = query_metadata_by_component(app, "FILTER", "2D Spectra", FALLBACK_NAME)
+            gratings = query_metadata_by_component(app, "GRATING", "2D Spectra", FALLBACK_NAME)
 
-    if np.all([isinstance(x, fits.HDUList) for x in data_obj]):
+            filters_gratings = [(f+'/'+g) for f, g in zip(filters, gratings)]
```

Comment:
I noticed when working on getting NIRCam to load that there is also some of this information in the "PUPIL" header keyword. It seems like the different instruments put different information in FILTER/GRATING/PUPIL - maybe this would be a good opportunity to get all three and display whichever ones are populated in their own separate columns. I think it should be as simple as adding `pupil = query_metadata...` and having three separate `add_to_table` later.

Reply:
This is a point I've been thinking of as well. We have historically combined Filter/Grating, but I've been thinking of whether it would be useful to split this out to separate columns in the table. (AKA, why are we even combining them in the first place?). This would also give room to include another Pupil field as well, and yes it would be as simple as what you describe above.

I hesitated on doing that here, as I wanted this PR to mainly keep the same functionality, but demonstrate the new infrastructure. I'd advocate a separate discussion with a PO/scientist to confirm whether this is what we want to do, and to a separate PR to make those changes, so that this infrastructure PR doesn't become even more bloated than it already is

Reply:
Alright, that would be a small follow-up PR anyway. Let me test this and make sure it looks good in-notebook, hopefully I'll approve shortly.

Reply:
If in reality I actually coded this to half as elegantly as I architected it, it should hopefully be a very small change